src/magic_folder/test/test_local_snapshot.py

```diff
@@ -0,0 +1,336 @@
+from __future__ import print_function
+
+import attr
+import os
+
+from hypothesis import (
+    given,
+)
+from hypothesis.strategies import (
+    binary,
+    lists,
+)
+
+from twisted.python.filepath import (
+    FilePath,
+)
+
+from twisted.internet import defer
+
+from testtools.matchers import (
+    Not,
+    Equals,
+    Always,
+    HasLength,
+    MatchesStructure,
+)
+from testtools.twistedsupport import (
+    succeeded,
+)
+from testtools import (
+    ExpectedException,
+)
+
+from eliot import (
+    Message,
+)
+
+from magic_folder.magic_folder import (
+    LocalSnapshotService,
+    LocalSnapshotCreator,
+)
+from magic_folder.snapshot import (
+    create_local_author,
+)
+from magic_folder.magicpath import (
+    path2magic,
+)
+from .. import magicfolderdb
+
+from .common import (
+    SyncTestCase,
+)
+from .strategies import (
+    path_segments,
+)
+
+@attr.s
+class MemorySnapshotCreator(object):
+    """
+    A way to test LocalSnapshotService with an in-memory database.
+    """
+    processed = attr.ib(default=attr.Factory(dict))
+    def process_item(self, path):
+        Message.log(
+            message_type=u"memory-snapshot-creator:process_item",
+            path=path.asTextMode("utf-8").path,
+        )
+        if path not in self.processed.keys():
```

Comment:
`x in dict` and `x in dict.keys()` are the same except `x in dict` is faster and uses less memory.